### PR TITLE
shattered-pixel-dungeon: add some mods/forks

### PR DESCRIPTION
**Pull request by fgaz**
## Description of changes

* split a generic builder from shattered-pixel-dungeon
* add a bunch of mods/forks of the game

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by gmemstr** [comment]
Result of `nixpkgs-review pr 254749` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>experienced-pixel-dungeon</li>
    <li>rat-king-adventure</li>
    <li>rkpd2</li>
    <li>shattered-pixel-dungeon</li>
    <li>shorter-pixel-dungeon</li>
    <li>summoning-pixel-dungeon</li>
  </ul>
</details>

**Comment by fgaz** [comment]
@gmemstr could you also have a look at the diff? :pray:

File: pkgs/games/shattered-pixel-dungeon/generic.nix [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    "jre"
+    "libpulseaudio"
+    "makeDesktopItem"
+    "copyDesktopItems"
+  ];
+
+  postPatch = ''
+    # disable gradle plugins with native code and their targets
+    perl -i.bak1 -pe "s#(^\s*id '.+' version '.+'$)#// \1#" build.gradle
+    perl -i.bak2 -pe "s#(.*)#// \1# if /^(buildscript|task portable|task nsis|task proguard|task tgz|task\(afterEclipseImport\)|launch4j|macAppBundle|buildRpm|buildDeb|shadowJar|robovm|git-version)/ ... /^}/" build.gradle
```
Comment (marcusramberg) [inline pkgs/games/shattered-pixel-dungeon/generic.nix:38]: Is there any reason to make backups of the replacements during this build process?

**Comment by marcusramberg** [comment]
Tried building on aarch64-linux but saw this error
```
error: hash mismatch in fixed-output derivation '/nix/store/m3wdrzxzmlvc4fq90ivl64w7aampaz2p-shorter-pixel-dungeon-deps-1.2.0.drv':
         specified: sha256-i4k5tdo07E1NJwywroaGvRjZ+/xrDp6ra+GTYwTB7uk=
            got:    sha256-QfAV6LYD6S/8ptaqqKSDtOe4kStwp6LJp8WVc3sH8yc=
error: 1 dependencies of derivation '/nix/store/pgbfs38k6rhsj7wgkwmp83aqk6xkjjn6-shorter-pixel-dungeon-1.2.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7grxmq10ynp5l0an503s4bj5cczmck38-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/254749

1 package failed to build:
shorter-pixel-dungeon
``` 

when running `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 254749"`

**Comment by fgaz** [comment]
A SNAPSHOT dependency made the FOD nondeterministic. Fixed.

File: pkgs/games/shattered-pixel-dungeon/generic.nix (hunk 1 above)
Comment (fgaz) [reply pkgs/games/shattered-pixel-dungeon/generic.nix:38]: I guess it makes it easier to inspect the build if something goes wrong, and it takes almost no resources anyway.

I actually copied this part from another gradle package :-P

**Comment by fgaz** [comment]
Found and replaced another SNAPSHOT. Should be fine now.

**Review by marcusramberg — APPROVED** [review]
lgtm, nix-review on linux seemed good.

**Review by marcusramberg — APPROVED** [review]

**Pull request merged**